File: 4800/4803.py
def PRINT(ans,z):
    if ans==0:print('Case {}: No trees.'.format(z))
    elif ans == 1:print('Case {}: There is one tree.'.format(z))
    else:print('Case {}: A forest of {} trees.'.format(z, ans))
def A(x):
    cnt = 1
    check[x]=1
    for nx in Tree[x]:
        if not check[nx]:
            cnt += A(nx)
    return cnt
def B(x):
    cnt = len(Tree[x])
    Pass[x]=1
    for nx in Tree[x]:
        if not Pass[nx]:
            cnt += B(nx)
    return cnt
from collections import deque
input=__import__('sys').stdin.readline
z=1
while z:
    n,m=map(int,input().split())
    if n==m==0:break
    check=[0]*n
    Pass=[0]*n
    Tree=[[] for _ in range(n)]
    ans=0
    for i in range(m):
        a,b=map(int,input().split())
        Tree[a-1].append(b-1)
        Tree[b-1].append(a-1)
    for i in range(n):
        if not check[i]:
            V,E=A(i),B(i)
            if V-1==E//2:ans+=1
    PRINT(ans,z)
    z+=1

File: 4800/4803.cpp
#include <bits/stdc++.h>
using namespace std;

class Tree{
public:
    int N;
    vector<int> visited;
    int arr[500][500] = {0};

    Tree():N(0){}
    Tree(int n):N(n){
        visited.resize(N,0);
    }
    void connect_node(int x, int y){
        arr[x][y] = 1;
        arr[y][x] = 1;
    }
    int find_tree(int a){
        queue<int> q;
        q.push(a);
        visited[a] = 1;
        while(!q.empty()){
            int x = q.front(); q.pop();
            for(int i=0; i<N; i++){
                if(arr[x][i]){
                    if(visited[i] > 0) return 0;
                    arr[x][i] = 0; arr[i][x] = 0;
                    q.push(i);
                    visited[i] = 1;
                }
            }
        }
        return 1;
    }
};

int main(){
    int n,m,a,b,z=1;
    while(1) {
        cin >> n >> m;
        if(n==0 && m==0) break;

        Tree T(n);
        for(int i=0; i<m; i++){
            cin >> a >> b;
            T.connect_node(a-1,b-1);
        }
        int ans = 0;
        for(int i=0; i<n; i++){
            if(!T.visited[i]) ans+=T.find_tree(i);
        }

        cout << "Case " << z <<": ";
        if(ans==0) cout << "No trees.";
        else if(ans==1) cout << "There is one tree.";
        else cout << "A forest of " << ans << " trees.";
        cout << endl;
        z++;
    }
}